core/arch/arm/kernel/smccc_call_a64.S

```diff
@@ -0,0 +1,25 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (C) 2021 Foundries.io Ltd
+ */
+
+#include <arm.h>
+#include <asm.S>
+#include <generated/asm-defines.h>
+#include <smccc.h>
+
+/*
+ * void smccc_smc(unsigned long a0, unsigned long a1, unsigned long a2,
+ *		  unsigned long a3, unsigned long a4, unsigned long a5,
+ *		  unsigned long a6, unsigned long a7, struct smccc_res *res)
+ * a0~a7: input arguments
+ * res: output arguments from register 0-3
+ */
+FUNC smccc_smc , :
+	smc	#0
+	ldr	x4, [sp]
+	/* Put results to res struct */
+	stp	x0, x1, [x4, #SMCCC_RES_X0_OFFS]
+	stp	x2, x3, [x4, #SMCCC_RES_X2_OFFS]
```

Comment:
Can't `sp` be used directly instead of `x4`?

Reply:
you mean in both lines?
well, not really, as `sp` value will change after pushing `x0`,`x1` values to stack on line 22.

if you mean on line 23 only, IMO I see no value for that

Reply:
I'd prefer to do that the same way as it's done in Linux kernel source code
https://github.com/torvalds/linux/blob/master/arch/arm64/kernel/smccc-call.S#L41

IMO Explicitly specifying the offsets (`SMCCC_RES_X0_OFFS` etc) makes that piece of code more readable.

Reply:
Actually we have a macro for that:
```
#include <arm64_macros.S>

store_xregs sp, SMCCC_RES_X0_OFFS, 0, 3
```
Then you can drop the definition of `SMCCC_RES_X2_OFFS`. BTW I suggest also dropping the `_OFFS` suffix, it is more or less implicit (check the names of other defines in `out/arm/core/include/generated/asm-defines.h`).

Reply:
Actually we have a macro for that:
```
#include <arm64_macros.S>

store_xregs sp, SMCCC_RES_X0_OFFS, 0, 3
```
Then you can drop the definition of `SMCCC_RES_X2_OFFS`. BTW I suggest also dropping the `_OFFS` suffix, it is more or less implicit (check the names of other defines in `out/arm/core/include/generated/asm-defines.h`).

Reply:
Interesting, but that doesn't work

```
diff --git a/core/arch/arm/kernel/asm-defines.c b/core/arch/arm/kernel/asm-defines.c
index 31c38244..b83c665b 100644
--- a/core/arch/arm/kernel/asm-defines.c
+++ b/core/arch/arm/kernel/asm-defines.c
@@ -125,7 +125,6 @@ DEFINES
               offsetof(struct boot_embdata, reloc_len));
 #ifdef ARM64
        /* struct smccc_res */
-       DEFINE(SMCCC_RES_X0_OFFS, offsetof(struct smccc_res, a0));
-       DEFINE(SMCCC_RES_X2_OFFS, offsetof(struct smccc_res, a2));
+       DEFINE(SMCCC_RES_X0, offsetof(struct smccc_res, a0));
 #endif
 }
diff --git a/core/arch/arm/kernel/smccc_call_a64.S b/core/arch/arm/kernel/smccc_call_a64.S
index 3b6604f9..351ec1c4 100644
--- a/core/arch/arm/kernel/smccc_call_a64.S
+++ b/core/arch/arm/kernel/smccc_call_a64.S
@@ -4,6 +4,7 @@
  */
 
 #include <arm.h>
+#include <arm64_macros.S>
 #include <asm.S>
 #include <generated/asm-defines.h>
 
@@ -18,7 +19,6 @@ FUNC smccc_smc , :
        smc     #0
        ldr     x4, [sp]
        /* Put results to res struct */
-       stp     x0, x1, [x4, #SMCCC_RES_X0_OFFS]
-       stp     x2, x3, [x4, #SMCCC_RES_X2_OFFS]
+       store_xregs sp, SMCCC_RES_X0, 0, 3
        ret
 END_FUNC smccc_smc

```

Maybe that's because of `sp` aligment restrictions?
> This mechanism is simple in principle: if the alignment restrictions of sp are inconvenient, just use another register as your stack pointer. General-purpose registers have no special alignment restrictions.

https://community.arm.com/developer/ip-products/processors/b/processors-ip-blog/posts/using-the-stack-in-aarch64-implementing-push-and-pop

Reply:
What doesn't work? Compilation or runtime? I get this disassembly:
```
0000000000000000 <smccc_smc>:
smccc_smc():
/home/jerome/work/optee_repo_qemu_v8/optee_os/core/arch/arm/kernel/smccc_call_a64.S:19
   0:   d4000003        smc     #0x0
/home/jerome/work/optee_repo_qemu_v8/optee_os/core/arch/arm/kernel/smccc_call_a64.S:21
   4:   a90007e0        stp     x0, x1, [sp]
   8:   a9010fe2        stp     x2, x3, [sp, #16]
/home/jerome/work/optee_repo_qemu_v8/optee_os/core/arch/arm/kernel/smccc_call_a64.S:22
   c:   d65f03c0        ret
```
Looks pretty good, doesn't it?

Reply:
@jforissier looks like this is the problem, as this works:
```
ldr     x4, [sp]                                                        
store_xregs x4, SMCCC_RES_X0, 0, 3  
```

Reply:
Oh, did I miss an indirection? ;-) My assembly skills are not as good as I'd like... LGTM

Reply:
runtime,  it hangs when trying to store value using directly the stack pointer as a base address